LibraryInstrumentationGuidance.md

```diff
@@ -0,0 +1,107 @@
+    finally
+    {
+        // stop activity if started
+        if (activity != null)
+             source.StopActivity(activity, new { Input = input, Output = output }); 
+    }
+```
+> ### *__TODO__ - provide a pointer to a document with more advanced instrumentation examples (WIP)* 
+
+### Payload
+
+When starting and stopping an activity it is possible to pass a payload object which will be available for the listener. This allows the listener to access raw, non-serialized data in real time. It can be useful for pulling out additional diagnostic information but also manipulating data as it is being processed (e.g. inject diagnostic context before an outbound call is made). Good examples of such payload are ```HttpRequestMessage``` or messages that are being passed through queues.
+
+Mind that payload is not preserved as part of Activity and is only available when activity is started/stopped so it is a good practice to specify all data that was passed to ```StartActivity()``` in ```StopActivity()``` as well.
+
+#### Payload format
+
+Diagnostic source event and activity start/stop API allows to specify only a single payload object, however in order to pass more data and allow future additions the recommendation is to use dynamic objects. Since these are .NET objects the names of particular properties should follow [standard .NET naming convention][DotNetPropertyNamingConvention]. 
+
+Here are some recommendations for typical payload property names:
+
+| Property name | Description |
+|:--------------|:-------------------|
+| `Endpoint` | The ```Uri``` of an endpoint the activity is for (e.g. target database, service) |
+| `PartitionKey` | The key/ID of the partition the activity is for |
+| `Status` | The ```TaskStatus``` of a completed asynchronous task |
+| `Exception` | The captured exception object |
+
+### Performance
+
+In order to avoid unnecessary overhead it is highly recommended to check if there is any listener for given activity. This can be done by calling 
+
+```C#
+    source.IsEnabled("Example.MyLibrary.MyOperation")
+```
+
+It is also possible to specify additional context payload when making that call to allow the listener to make a more informed decision (e.g. listeners may only be interested in activities for certain endpoint or partition). However, since this call is performed for every operation, it is NOT recommended to build a dynamic payload object as described above. Instead the raw input objects should be specified directly in the call - the Diagnostic Source API allows to specify up to 2 payload objects:
+
+```C#
+    source.IsEnabled("Example.MyLibrary.MyOperation", input1, input2)
+```
+
+For more detailed performance considerations please refer to [Diagnostic Source][DiagnosticSourceGuide] and [Activity][ActivityGuide] guides.
+
+### Tags
+
+Activities can have additional tracing information in tags. Tags are meant to be easily consumable and are expected to be logged with the activity without any processing. 
```

Comment:
Please mention that Tags should be set (if any) as early as possible - e.g. it does not make sense to set them right after calling StartActivity, they should be set before that (or before StopActivity).

I  think it's also worth mentioning what should be added and performance implications of that:  Tags are intended to be added along the telemetry blindly, so they should be used wisely (just set a few) and are completely optional. In some sense, they substitute event payloads and give tracing system essential information about the operation (while payloads give full context).

Reply:
Thanks, updated this paragraph, please review